Feedback UI documentation (#57)
bucketco/bucket-tracking-sdk: FEEDBACK.md

# Bucket Feedback UI

> The Bucket SDK includes a UI you can use to collect feedback from user about particular features.

Author: ```suggestion
The Bucket SDK includes a user interface (UI) you can use to collect feedback from user about particular features.
```

## Global feedback configuration

> The Bucket SDK feedback UI is configured with reasonable defaults of an english language, positioning itself as a [dialog](#dialog) in the lower right-hand corner of the viewport and with a [light-mode theme](#custom-styling).

Author: ```suggestion
The Bucket SDK feedback UI is configured with reasonable defaults, positioning itself as a [dialog](#dialog) in the lower right-hand corner of the viewport, displayed in english, and with a [light-mode theme](#custom-styling).
```

These settings can be overwritten when initializing the Bucket SDK:

```javascript
bucket.init("bucket-tracking-key", {
  feedback: {
    ui: {
      position: POSITION_CONFIG, // See positioning section
      translations: TRANSLATION_KEYS, // See internationalization section

      // Enable automated fedback collection. Default: `false`
      automaticPrompting: boolean,

      /**
       * Do your own feedback prompt handling or override
       * default settings at runtime.
       */
      promptHandler: (promptMessage, handlers) => {
        // See automated feedback collection section
      },
    },
  },
});
```

See also:

- [Positioning and behavior](#positioning-and-behavior) for the position option.
- [Static language configuration](#static-language-configuration) if you want to translate the feedback UI.
- [Automated feedback collection](#automated-feedback-collection) to override default configuration.

## Manual feedback collection

> You can open up the feedback collection UI by calling `bucket.requestFeedback(options)` with the relevant options. This is great is you are interested in manually controlling the collection of feedback from your user, but want to use the convenience of the Bucket feedback UI to reduce the amount of code you need to maintain.

Author: The suggestion, filtered for [words to avoid in educational writing](https://css-tricks.com/words-avoid-educational-writing/):

```suggestion
To open up the feedback collection interface, call `bucket.requestFeedback(options)` with the appropriate options. This approach is particularly beneficial if you wish to retain manual control over feedback collection from your users while leveraging the convenience of the Bucket feedback UI to reduce the amount of code you need to maintain.
```

> Examples of this could be if you want to put a `give us feedback`-button on your page, or pop up the feedback UI at the end of specific user flows in your app.

Author: ```suggestion
Examples of this could be if you want the click of a `give us feedback`-button or the end of a specific user flow, to trigger a pop-up displaying the feedback user interface.
```

### bucket.requestFeeback() options

Minimal usage with defaults:

```javascript
bucket.requestFeedback({
  featureId: "bucket-feature-id",
  title: "How satisfied are you with file uploads?",
});
```

All options:

> ```javascript
> bucket.requestFeedback({
>   featureId: "bucket-feature-id", // required
>   userId: "your-user-id", // optional if user persistence is enabled (default in browsers),
>   companyId: "users-company-or-account-id", // optional
>   title: "How satisfied are you with file uploads?"
> 
>   position: POSITION_CONFIG, // See positioning section
>   translations: TRANSLATION_KEYS // See internationalization section
> 
>   // Trigger side effects with the collected data,
>   // for example posting it back into your own CRM
>   onAfterSubmit: (feedback) => {
>     storeFeedbackInCRM({
>       score: feedback.score,
>       comment: feedback.comment
>     })
>   }
> })
> ```

Author: ```suggestion
```javascript
bucket.requestFeedback({
  featureId: "bucket-feature-id", // [Required]
  userId: "your-user-id",  // [Optional] if user persistence is enabled (default in browsers),
  companyId: "users-company-or-account-id", // [Optional]
  title: "How satisfied are you with file uploads?" // [Optional]

  position: POSITION_CONFIG, // [Optional] see the positioning section
  translations: TRANSLATION_KEYS // [Optional] see the internationalization section

  // [Optional] trigger side effects with the collected data,
  // for example sending the feedback to your own CRM
  onAfterSubmit: (feedback) => {
    storeFeedbackInCRM({
      score: feedback.score,
      comment: feedback.comment
    })
  }
})
```
```

See also:

- [Positioning and behavior](#positioning-and-behavior) for the position option.
- [Runtime language configuration](#runtime-language-configuration) if you want to translate the feedback UI.

## Automated feedback collection

> When automated feedback collection is enabled the Bucket SDK will stay in contact with the Bucket service. When a user triggers an event used in a feature, and the user is eligible for being prompted for feedback, the Bucket service will push an event to the SDK instance. By default, this event will open up the Bucket feedback UI, but you have some abilities to intercept this event and add your own overrides to it.

Author: ```suggestion
When automated feedback collection is enabled, the Bucket SDK will open and maintain a connection to the Bucket service. When a user triggers an event tracked by a feature and is eligible to be prompted for feedback, the Bucket service will send a request to the SDK instance. By default, this request will open up the Bucket feedback interface in the user's browser, but you can intercept the request and override this behavior.
```

```javascript
bucket.init({
  feedback: {
    automaticPrompting: true,
    promptHandler: (promptMessage, handlers) => {
      // Pass your overrides here. Everything is optional
      handlers.openFeedbackForm({
        title: promptMessage.question,

        position: POSITION_CONFIG, // See positioning section
        translations: TRANSLATION_KEYS, // See internationalization section

        // Trigger side effects with the collected data,
        // for example posting it back into your own CRM
        onAfterSubmit: (feedback) => {
          storeFeedbackInCRM({
            score: feedback.score,
            comment: feedback.comment,
          });
        },
      });
    },
  },
});
```

See also:

- [Positioning and behavior](#positioning-and-behavior) for the position option.
- [Runtime language configuration](#runtime-language-configuration) if you want to translate the feedback UI.
- [Use your own UI to collect feedback](#using-your-own-ui-to-collect-feedback) if the feedback UI doesn't match your design.

## Positioning and behavior

The feedback UI can be configured to be placed and behave in 3 different ways:

### Positioning configuration

#### Modal

> A modal overlay with a backdrop that blocks interaction with the remaining page. Can be dismissed with `<ESC>` or a close button. It is always centered on the page, making it the primary thing the user needs to focus on.

Reviewer: I suggest:

A modal overlay with a backdrop that blocks interaction with the underlying page. It can be dismissed with the keyboard shortcut `<ESC>` or the dedicated close button in the top right corner. It is always centered on the page, capturing focus, and making it the primary interface the user needs to interact with.

Author: ```suggestion
A modal overlay with a backdrop that blocks interaction with the underlying page. It can be dismissed with the keyboard shortcut `<ESC>` or the dedicated close button in the top right corner. It is always centered on the page, capturing focus, and making it the primary interface the user needs to interact with.
```

> Using a modal is the strongest possible push for feedback. You are interrupting the users normal flow, which can cause annoyance. A good use case fora modal is when the user finishes a linear flow that they don"t use often, for example setting up an account.

Reviewer: I suggest:

Using a modal is the strongest possible push for feedback. You are interrupting the user's normal flow, which can cause annoyance. A good use-case for the modal is when the user finishes a linear flow that they don't perform often, for example setting up a new account.

Author: ```suggestion
Using a modal is the strongest possible push for feedback. You are interrupting the user's normal flow, which can cause annoyance. A good use-case for the modal is when the user finishes a linear flow that they don't perform often, for example setting up a new account.
```

```javascript
position: {
  type: "MODAL";
}
```

#### Dialog

> A dialog that appears in a corner of the viewport, but lets the user continue with their interaction with the page. Can be dismissed with a close button, but also automatically disappears after a relatively shot time period if the user does not interact with it.

Reviewer: I suggest:

A dialog that appears in a specified corner of the user's browser, without limiting the user's interaction with the rest of the page. It can be dismissed with the dedicated close button, but will automatically disappear after a short time period if the user does not interact with it.

Author: ```suggestion
A dialog that appears in a specified corner of the user's browser, without limiting the user's interaction with the rest of the page. It can be dismissed with the dedicated close button, but will automatically disappear after a short time period if the user does not interact with it.
```

Author: applying suggestion, but keeping the technical lingo of `viewport`

> Using a dialog is a soft push for feedback. It lets the user continue with their work with a minimal amount of intrusion. The user can opt in to responding, but is not required to. A good use case for this behavior is when a user used a feature where the expected outcome is predictable, possibly because they have used it multiple times before. For example: Uploading a file, switching to a different view of a visualization, visiting a specific page or manipulating some data.

Reviewer: I suggest:

Using a dialog is a soft push for feedback. It lets the user continue their work with a minimal amount of intrusion. The user can opt-in to respond but is not required to. A good use case for this behavior is when a user uses a feature where the expected outcome is predictable, possibly because they have used it multiple times before. For example: Uploading a file, switching to a different view of a visualization, visiting a specific page, or manipulating some data.

Reviewer: > A good use case for this behavior is when a user uses a feature where the expected outcome is predictable...

I am a little unsure as to why predictability is beneficial here. Would appreciate an explanation 😄

Author: > I am a little unsure as to why predictability is beneficial here. Would appreciate an explanation 😄

Predictability in terms of the user knows what to do next, possibly because they've made the interaction before, or the interface they are interacting with makes future affordances clear. In those cases, you can assume the users knows what to do next, and thus you likely want to avoid interrupting them

> The default feedback UI behavior is a dialog placed at the bottom right of the users viewport.

Reviewer: I suggest:

The default feedback interface behavior is a dialog placed in the bottom right corner of the user's browser.

Author: ```suggestion
The default feedback interface behavior is a dialog placed in the bottom right corner of the user's browser.
```

```javascript
position: {
  type: "DIALOG",
  placement: "top-left" | "top-right" | "bottom-left" | "bottom-right"
}
```

#### Popover

> A popover that is anchored relative to a DOM-element (typically a button). Can be dismissed with an outside click or pressing a close button.

Reviewer: I suggest:

A popover that is anchored relative to a DOM-element (typically a button). It can be dismissed by clicking outside the popover or by pressing the dedicated close button.

Author: ```suggestion
A popover that is anchored relative to a DOM-element (typically a button). It can be dismissed by clicking outside the popover or by pressing the dedicated close button.
```

You can use the popover mode to implement your own button to collect feedback manually.

```javascript
position: {
  type: "POPOVER",
  anchor: DOMElement
}
```

Popover feedback button example:

```html
<button id="feedbackButton">Tell us what you think</button>
<script>
  const button = document.getElementById("feedbackButton");
  button.addEventListener("click", (e) => {
    bucket.requestFeedback({
      featureId: "bucket-feature-id",
      userId: "your-user-id",
      title: "How do you like the popover?",
      position: {
        type: "POPOVER",
        anchor: e.currentTarget,
      },
    });
  });
</script>
```

## Internationalization (i18n)

> The Feedback UI currently only comes in English. You can supply your own translation keys that will be used insude the UI by passing an object with your own translation keys.

Reviewer: I suggest:

By default, the feedback interface is written in English. However, you can supply your own translations by passing an object to the options to either or both of the `bucket.init(options)` or `bucket.requestFeedback(options)` calls. These translations will replace the English ones used by the feedback interface. See examples below.

Author: ```suggestion
By default, the feedback interface is written in English. However, you can supply your own translations by passing an object to the options to either or both of the `bucket.init(options)` or `bucket.requestFeedback(options)` calls. These translations will replace the English ones used by the feedback interface. See examples below.
```

See [default english localization keys](./src/feedback/config/defaultTranslations.tsx) for a reference of what translation keys can be supplied.

### Static language configuration

If you know the language at page load, you can configure your translation keys while initializing the Bucket SDK:

```javascript
bucket.init("my-tracking-key", {
  feedback: {
    ui: {
      translations: {
        // Your translation keys
      },
    },
  },
});
```

### Runtime language configuration

> If you only know the users language after the page has loaded, you can provide translation keys in the two cases where you open the feedback UI:

Reviewer: I suggest:

If you only know the user's language after the page has loaded, you can provide translations to either the `bucket.requestFeedback(options)` call or the `promptHandler` option before the feedback interface opens. See examples below.

### Manual feedback collection

```javascript
bucket.requestFeedback({
  ... // Other options
  translations: {
    // your translation keys
  }
})
```

### Automated feedback collection

When you are collecting feedback through the Bucket automation, you can intercept the default prompt handling and override the defaults.

> If you set the prompt question in the Bucket app to be one of your own translation keys, you can even get a translated version of the question you want to ask your customer in the feedback UI.

Reviewer: I suggest:

If you set the prompt question in the Bucket app to be one of your own translations, you even get a translated version of the question you want to ask your customer in the feedback interface.

Author: ```suggestion
If you set the prompt question in the Bucket app to be one of your own translations, you even get a translated version of the question you want to ask your customer in the feedback interface.
```

Author: Discarding suggestion. In this case `translation key` is actually what they want to be setting

```javascript
bucket.init({
  feedback: {
    promptHandler: (message, handlers) => {
      const translatedQuestion =
        i18nLookup[message.question] ?? message.question;
      handlers.openFeedbackForm({
        title: translatedQuestion,
        translations: {
          // your static translation keys
        },
      });
    },
  },
});
```

## Custom styling

> You can adapt parts of the looks of the Bucket feedback UI by setting CSS custom properties on your CSS `:root`-scope.

Reviewer: I suggest:

You can adapt parts of the look of the Bucket feedback interface by supplying CSS custom properties to your page in your CSS `:root`-scope.

Author: ```suggestion
You can adapt parts of the look of the Bucket feedback interface by supplying CSS custom properties to your page in your CSS `:root`-scope.
```

Examples of custom styling can be found in our [development example stylesheet](./dev/index.css).

> ## Using your own UI to collect feedback

Reviewer: I suggest:

## Using your own interface to collect feedback

Author: Adding this change will require updates of all section links

```suggestion
## Using your own interface to collect feedback
```

Author: Discarding this suggestion, since we stick with the UI term

> You might have very strict design guidelines in your app, and maybe the Bucket feedback UI doesn"t quite work for you.

Reviewer: You may have very strict design guidelines for your app and maybe the Bucket feedback UI doesn't quite work for you.

Author: ```suggestion
You may have very strict design guidelines for your app and maybe the Bucket feedback UI doesn't quite work for you.
```

> In this case, you can actually implement your own feedback collection mechanism, which follows your own design guidelines.

Reviewer: I suggest:

In this case, you can implement your own feedback collection mechanism, which follows your own design guidelines.

Author: ```suggestion
In this case, you can implement your own feedback collection mechanism, which follows your own design guidelines.
```

This is the data type you need to collect:

```typescript
{
  /** Customer satisfaction score */
  score?: 1 | 2 | 3 | 4 | 5,
  comment?: string
}
```

Either `score` or `comment` must be defined in order to pass validation in the Bucket tracking API.

### Manual feedback collection

Examples of a HTML-form that collects the relevant data can be found in [feedback.html](./example/feedback/feedback.html) and [feedback.jsx](./example/feedback/feedback.jsx).

> Once you have the feedback data, pass it along to `bucket.feedback()`:

Reviewer: I suggest:

Once you have collected the feedback data, pass it along to `bucket.feedback()`:

Author: ```suggestion
Once you have collected the feedback data, pass it along to `bucket.feedback()`:
```

```javascript
bucket.feedback({
  featureId: "bucket-feature-id",
  userId: "your-user-id",
  score: 5,
  comment: "Best thing I"ve ever tried!",
});
```

### Intercepting automated feedback collection events

> When using automated feedback collection, Bucket will some times send a feedback prompt message to your users instance of the Bucket SDK. This will result in the feedback UI being opened.

Reviewer: I suggest:

When using automated feedback collection, the Bucket service will, when specified, send a feedback prompt message to your user's instance of the Bucket SDK. This will result in the feedback interface being opened.

Author: ```suggestion
When using automated feedback collection, the Bucket service will, when specified, send a feedback prompt message to your user's instance of the Bucket SDK. This will result in the feedback interface being opened.
```
